File: Leetcode/tree/257_BinaryTreePaths/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    void dfs(TreeNode* root, string s, vector<string>& res){
        if(!root->left && !root->right){
            res.push_back(s);
            return;
        }
        if(root->left!=nullptr) dfs(root->left, s+"->"+to_string(root->left->val), res);
        if(root->right!=nullptr) dfs(root->right, s+"->"+to_string(root->right->val), res);
    }
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> res;
        if(root==nullptr) return res;
        dfs(root, to_string(root->val), res);
        return res;
    }
};

File: Leetcode/tree/257_BinaryTreePaths/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]
        """
        if not root:
            return []
        res = []
        self.dfs(root, "", res)
        return res

    def dfs(self, root, path, res):
        if not root.left and not root.right:
            path += str(root.val)
            res.append(path)
        if root.left:
            self.dfs(root.left,  path+str(root.val)+"->", res)
        if root.right:
            self.dfs(root.right, path+str(root.val)+"->", res)
